hwcomposer/hwcomposer.cpp

```diff
@@ -319,6 +319,21 @@ static void apply_surface_damage(hwc_layer_1 *hwc_layer, surface_context &surfac
     });
 }
 
+static void apply_opaque_region(waydroid_hwc_composer_device_1 *pdev, hwc_layer_1 *hwc_layer, surface_context &surface_context) {
+    const auto& scale = pdev->display->scale;
+    const auto& visible_regions = hwc_layer->visibleRegionScreen;
+    auto region = wl_compositor_create_region(pdev->display->compositor);
+    std::for_each(visible_regions.rects, visible_regions.rects + visible_regions.numRects, [&](const auto& rect){
+        int32_t x = static_cast<int32_t>(rect.left / scale);
+        int32_t y = static_cast<int32_t>(rect.top / scale);
+        int32_t width = static_cast<int32_t>(ceil((rect.right - rect.left) / scale));
+        int32_t height = static_cast<int32_t>(ceil((rect.bottom - rect.top) / scale));
+        wl_region_add(region, x, y, width, height);
+    });
+    wl_surface_set_opaque_region(surface_context.surface, region);
```

Comment:
I'm worried that visibleRegionScreen doesn't really map to opaque regions. What about translucent shadows in multi-windows mode?

Reply:
Sounds reasonable. I will drop it for now